### PR TITLE
[Site Isolation] Make `findString:` correctly cycle through matches and update selection

### PR DESCRIPTION
**Pull request by charliewolfe**
#### bf1345fe689b99d16d63b52702777db6c79e1717
<pre>
[Site Isolation] Make `findString:` correctly cycle through matches and update selection
<a href="https://bugs.webkit.org/show_bug.cgi?id=267908">https://bugs.webkit.org/show_bug.cgi?id=267908</a>
<a href="https://rdar.apple.com/121411740">rdar://121411740</a>

Reviewed by Alex Christensen.

This patch adds initial support for the UI process to coordinate `findString:` across multiple web
processes.

To make this work with site isolation, I needed to:
- Search all web processes without setting selection and have them reply with the identifier of the frame
  containing the string and if the last match was cycled.
- Traverse the frame tree in the UI process to determine which frame is closest to the focused frame.
- Consider when all matches have been cycled in a process and when all matches have been cycled in the
  page.
- Change logic in the web process to consider an out-of-process focused or main frame during frame
  traversal.

This assumes that all web processes will respond. There still needs to be work to make this robust
against unresponsive web processes. More details below.

* LayoutTests/platform/ios/fast/scrolling/find-text-in-overflow-node.html:
* LayoutTests/platform/ios/fast/scrolling/find-text-in-subframe.html:
Wait for a presentation update on some layout tests using findString.

* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::findString):
* Source/WebCore/editing/EditorCommand.cpp:
(WebCore::executeFindString):
* Source/WebCore/editing/FindOptions.h:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::find const):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::findString):
* Source/WebCore/page/Page.h:
`FocusController::focusedOrMainFrame()` will return the root frame if the focused and main frame are both
remote. This means that if there is a local frame containing a match between the root frame and the
out-of-process focused frame, any local frame following the focused remote frame will not be searched. To
fix this, we should always begin the frame traversal at the focused or main frame, even if they are
out-of-process.

We should only update the focused frame if selection is being set.

Also return the identifier of the frame containing the match. std::nullopt indicates no match.

* Source/WebKit/Shared/WebFindOptions.cpp:
(WebKit::core):
* Source/WebKit/Shared/WebFindOptions.h:
* Source/WebKit/Shared/WebFindOptions.serialization.in:
Add a `SetSelection` value to `FindOptions` to indicate when we only want to search the frame.

* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/FindStringCallbackAggregator.cpp: Added.
(WebKit::FindStringCallbackAggregator::create):
(WebKit::FindStringCallbackAggregator::foundString):
This function is called in response to each FindString IPC message that does not set selection.

(WebKit::FindStringCallbackAggregator::~FindStringCallbackAggregator):
Traverse the frame tree to find the frame closest to the focused frame containing a match. If findString
has cycled through all matches in the process but not the page the frame should not be considered. If the
frame is in a different process than the frame which previously had focus, send an IPC message to clear
selection from the previously focused frame.

(WebKit::FindStringCallbackAggregator::FindStringCallbackAggregator):

* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::didCreateSubframe):
(WebKit::WebFrameProxy::traverseNext const):
(WebKit::WebFrameProxy::traversePrevious):
(WebKit::WebFrameProxy::isDescendantOf const):
(WebKit::WebFrameProxy::deepLastChild):
(WebKit::WebFrameProxy::firstChild const):
(WebKit::WebFrameProxy::lastChild const):
* Source/WebKit/UIProcess/WebFrameProxy.h:
(WebKit::WebFrameProxy::parentFrame const):
(WebKit::WebFrameProxy::nextSibling const):
(WebKit::WebFrameProxy::previousSibling const):
(WebKit::WebFrameProxy::parentFrame): Deleted.
Added member variables and functions for traversing the frame tree, designed to behave similarly to the
frame traversal functions in `FrameTree`.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::findString):
* Source/WebKit/UIProcess/WebPageProxy.h:

* Source/WebKit/UIProcess/WebProcessPool.cpp:
* Source/WebKit/UIProcess/WebProcessPool.h:
There were two `audibleActivityClearDelay`` variables that started conflicting due to the unified sources
shift. I moved this one to fix it.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

* Source/WebKit/WebProcess/WebPage/FindController.cpp:
* Source/WebKit/WebProcess/WebPage/FindController.h:
(WebKit::FindController::updateFindUIAfterPageScroll):
Selection should not be cleared on searches where selection cannot be set.

(WebKit::FindController::findString):
Pass the frame identifier and didWrap to the completion handler instead of whether a match was found.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::findStringFromInjectedBundle):
(WebKit::WebPage::findString):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::siteIsolatedViewAndDelegate):
(TestWebKitAPI::TEST):
Add tests to verify `findString:` correctly updates selection on pages with various frame orderings.

* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[TestWKWebView selectionRangeHasStartOffset:endOffset:inFrame:])
Add a function similar to the existing `selectionRangeHasStartOffset:endOffset:`, used to check selection
within a frame.

Canonical link: <a href="https://commits.webkit.org/273517@main">https://commits.webkit.org/273517@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/dd6c405a02e7200d9e8b23bf852a078f51dc2342

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35585 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14526 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37722 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38326 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32062 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36783 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16909 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11552 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30867 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36138 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12273 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31668 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10780 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10796 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31806 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39571 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32336 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32142 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36750 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10980 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8867 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34815 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12692 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/31466 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/8141 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11489 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4620 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11760 "Built successfully") | | |

### CODE REVIEW

File: Source/WebKit/UIProcess/WebFrameProxy.h [hunk 1]
```diff
@@ -191,6 +204,8 @@ class WebFrameProxy : public API::ObjectImpl<API::Object::Type::Frame>, public C
     WebCore::FrameIdentifier m_frameID;
     ListHashSet<Ref<WebFrameProxy>> m_childFrames;
     WeakPtr<WebFrameProxy> m_parentFrame;
+    WeakPtr<WebFrameProxy> m_nextSibling;
+    WeakPtr<WebFrameProxy> m_previousSibling;
```
Comment (achristensen07) [inline Source/WebKit/UIProcess/WebFrameProxy.h]: I feel like these are redundant with having a ListHashSet of children.  To find the next sibling, you go to the parent's children, find(*this) gives you an iterator, and you can check if it's the end and increment if not.  I guess ListHashSet needs an rfind that gives you a reverse_iterator.

File: Source/WebKit/UIProcess/WebFrameProxy.h [hunk 2]
```diff
@@ -171,11 +173,22 @@ class WebFrameProxy : public API::ObjectImpl<API::Object::Type::Frame>, public C
...
+    WebFrameProxy* traverseNext(CanWrap, DidWrap* = nullptr) const;
+    WebFrameProxy* traversePrevious(CanWrap, DidWrap* = nullptr);
+
 private:
     WebFrameProxy(WebPageProxy&, WebProcessProxy&, WebCore::FrameIdentifier);
 
     std::optional<WebCore::PageIdentifier> pageIdentifier() const;
 
+    bool isDescendantOf(const WebFrameProxy* ancestor) const;
+    WebFrameProxy* deepLastChild();
```
Comment (achristensen07) [inline Source/WebKit/UIProcess/WebFrameProxy.h]: Can these return RefPtr?

File: Source/WebKit/UIProcess/WebFrameProxy.h [hunk 3]
```diff
@@ -171,11 +173,22 @@ class WebFrameProxy : public API::ObjectImpl<API::Object::Type::Frame>, public C
 
     bool isFocused() const;
 
+    WebFrameProxy* traverseNext(const WebFrameProxy* stayWithin = nullptr) const;
+    WebFrameProxy* traverseNext(CanWrap, DidWrap* = nullptr) const;
```
Comment (achristensen07) [inline Source/WebKit/UIProcess/WebFrameProxy.h]: Instead of having an optional out-parameter, let's return a struct like this:
struct TraversalResult {
    RefPtr<WebFrameProxy> frame;
    DidWrap didWrap;
};
Out-parameters are kind of an anti-pattern.

File: Source/WebKit/UIProcess/WebFrameProxy.h [hunk 4]
```diff
@@ -158,7 +160,7 @@ class WebFrameProxy : public API::ObjectImpl<API::Object::Type::Frame>, public C
     void getFrameInfo(CompletionHandler<void(FrameTreeNodeData&&)>&&);
     FrameTreeCreationParameters frameTreeCreationParameters() const;
 
-    WebFrameProxy* parentFrame() { return m_parentFrame.get(); }
+    WebFrameProxy* parentFrame() const { return m_parentFrame.get(); }
```
Comment (achristensen07) [inline Source/WebKit/UIProcess/WebFrameProxy.h:163]: Can we make this return a RefPtr while we're here?

File: Source/WebKit/WebKit.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -18943,7 +18949,6 @@
 			isa = PBXTargetDependency;
 			platformFilters = (
 				ios,
-				xros,
```
Comment (achristensen07) [inline Source/WebKit/WebKit.xcodeproj/project.pbxproj]: These changes look unrelated.  Let's undo them even though Xcode probably did this automatically.

File: Source/WebKit/UIProcess/WebFrameProxy.h (hunk 1 above)
Comment (charliewolfe) [reply Source/WebKit/UIProcess/WebFrameProxy.h]: FrameTree in WebCore stores sibling frames as member variables, so I was just following the pattern. But I guess we don't need to, I'll change it.

File: Source/WebKit/UIProcess/WebFrameProxy.h (hunk 3 above)
Comment (charliewolfe) [reply Source/WebKit/UIProcess/WebFrameProxy.h]: Similar here, I was following the pattern from FrameTree. But I will change this.

File: Source/WebKit/WebKit.xcodeproj/project.pbxproj (hunk 5 above)
Comment (charliewolfe) [reply Source/WebKit/WebKit.xcodeproj/project.pbxproj]: Whoops, will remove these.

File: Source/WebKit/UIProcess/WebProcessPool.cpp [hunk 6]
```diff
@@ -167,8 +167,6 @@ constexpr Seconds resetGPUProcessCrashCountDelay { 30_s };
 constexpr unsigned maximumGPUProcessRelaunchAttemptsBeforeKillingWebProcesses { 2 };
 #endif
 
-static constexpr Seconds audibleActivityClearDelay = 5_s;
```
Comment (pxlcoder) [inline Source/WebKit/UIProcess/WebProcessPool.cpp:170]: This change looks unrelated?

File: Source/WebKit/WebProcess/WebPage/WebPage.cpp [hunk 7]
```diff
@@ -5319,7 +5319,7 @@ void WebPage::setTextIndicator(const WebCore::TextIndicatorData& indicatorData)
 
 bool WebPage::findStringFromInjectedBundle(const String& target, OptionSet<FindOptions> options)
 {
-    return m_page->findString(target, core(options));
+    return !!m_page->findString(target, core(options));
```
Comment (pxlcoder) [inline Source/WebKit/WebProcess/WebPage/WebPage.cpp]: I feel that `.has_value()` is more readable.

File: Source/WebKit/WebProcess/WebPage/FindController.cpp [hunk 8]
```diff
@@ -274,13 +274,17 @@ void FindController::findString(const String& string, OptionSet<FindOptions> opt
...
     DidWrap didWrap = DidWrap::No;
 #if ENABLE(PDF_PLUGIN)
     if (pluginView)
         found = pluginView->findString(string, coreOptions, maxMatchCount);
     else
 #endif
-        found = m_webPage->corePage()->findString(string, coreOptions, &didWrap);
+    {
+        idOfFrameContainingString = m_webPage->corePage()->findString(string, coreOptions, &didWrap);
+        found = !!idOfFrameContainingString;
```
Comment (pxlcoder) [inline Source/WebKit/WebProcess/WebPage/FindController.cpp]: Same comment re: `has_value()`.

File: Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm [hunk 9]
```diff
@@ -1731,6 +1731,346 @@ HTTPServer server({
...
+        { { { 0, 11 }, { 0, 0 } } }
+    } };
+    for (auto& offsets : selectionOffsetsForFrames)
+        findStringAndValidateResults(offsets);
+}
+
+TEST(SiteIsolation, FindStringSelectionSameOriginFrameBeforeWrap)
+{
+    auto mainframeHTML = "<p>Hello world</p>"
+    "<iframe src='https://domain2.com/subframe'></iframe>"_s;
```
Comment (pxlcoder) [inline Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm]: Nit: I think we would normally indent these lines in a multiline literal. Same comment applies throughout the changes in file.

**Review by pxlcoder — COMMENTED** [review]
Nice! This is cleaner than I expected, and the test coverage looks really good.

File: Source/WebKit/UIProcess/WebFrameProxy.h (hunk 4 above)
Comment (charliewolfe) [reply Source/WebKit/UIProcess/WebFrameProxy.h:163]: I would need to change several call sites from `auto*` -> `RefPtr`. Seems like it should be a separate change?

File: Source/WebKit/UIProcess/WebProcessPool.cpp (hunk 6 above)
Comment (charliewolfe) [reply Source/WebKit/UIProcess/WebProcessPool.cpp:170]: There were two variables with this name in the global WebKit namespace and the unified sources shift made them conflict; so I moved this one.

File: Source/WebKit/UIProcess/WebFrameProxy.h (hunk 4 above)
Comment (achristensen07) [reply Source/WebKit/UIProcess/WebFrameProxy.h:163]: yes

File: Source/WebKit/UIProcess/WebFrameProxy.h [hunk 10]
```diff
@@ -171,11 +173,26 @@ class WebFrameProxy : public API::ObjectImpl<API::Object::Type::Frame>, public C
 
     bool isFocused() const;
 
+    struct TraversalResult {
+        RefPtr<WebFrameProxy> frame;
+        DidWrap didWrap;
+    };
+    TraversalResult traverseNext(const WebFrameProxy* stayWithin = nullptr) const;
```
Comment (achristensen07) [inline Source/WebKit/UIProcess/WebFrameProxy.h]: I don't think stayWithin is ever non-null.  Let's remove it.

File: Source/WebKit/UIProcess/WebFrameProxy.h [hunk 11]
```diff
@@ -171,11 +173,26 @@ class WebFrameProxy : public API::ObjectImpl<API::Object::Type::Frame>, public C
 
     bool isFocused() const;
 
+    struct TraversalResult {
+        RefPtr<WebFrameProxy> frame;
+        DidWrap didWrap;
```
Comment (achristensen07) [inline Source/WebKit/UIProcess/WebFrameProxy.h]: This needs a default initializer to DidWrap::No.  Otherwise it's uninitialized memory in a few places where you make a TraversalResult with { }.

File: Source/WebKit/UIProcess/WebFrameProxy.cpp [hunk 12]
```diff
@@ -557,6 +557,125 @@ std::optional<WebCore::PageIdentifier> WebFrameProxy::webPageIDInCurrentProcess(
...
+        return { previousSibling->deepLastChild(), DidWrap::No };
+    if (RefPtr parent = parentFrame())
+        return { parent, DidWrap::No };
+
+    if (canWrap == CanWrap::Yes)
+        return { deepLastChild(), DidWrap::Yes };
+    return { };
+}
+
+bool WebFrameProxy::isDescendantOf(const WebFrameProxy* ancestor) const
```
Comment (achristensen07) [inline Source/WebKit/UIProcess/WebFrameProxy.cpp]: This is never called.  Please remove.

File: Source/WebKit/UIProcess/WebFrameProxy.cpp [hunk 13]
```diff
@@ -557,6 +557,125 @@ std::optional<WebCore::PageIdentifier> WebFrameProxy::webPageIDInCurrentProcess(
...
+RefPtr<WebFrameProxy> WebFrameProxy::previousSibling() const
+{
+    if (!m_parentFrame)
+        return nullptr;
+
+    if (m_parentFrame->m_childFrames.first().ptr() == this)
+        return nullptr;
+
+    auto it = m_parentFrame->m_childFrames.find(this);
+    if (it == m_childFrames.end()) {
```
Comment (achristensen07) [inline Source/WebKit/UIProcess/WebFrameProxy.cpp:654]: This should check if it == m_childFrames.begin() and return nullptr if it is.

File: Source/WebKit/UIProcess/WebFrameProxy.cpp [hunk 14]
```diff
@@ -557,6 +557,125 @@ std::optional<WebCore::PageIdentifier> WebFrameProxy::webPageIDInCurrentProcess(
...
+RefPtr<WebFrameProxy> WebFrameProxy::nextSibling() const
+{
+    if (!m_parentFrame)
+        return nullptr;
+
+    if (m_parentFrame->m_childFrames.last().ptr() == this)
+        return nullptr;
+
+    auto it = m_parentFrame->m_childFrames.find(this);
+    if (it == m_childFrames.end()) {
```
Comment (achristensen07) [inline Source/WebKit/UIProcess/WebFrameProxy.cpp:638]: Before incrementing we should just assert that we are not at the end.  After incrementing we need a check to see if we are at the end, and return nullptr if we are.

File: Source/WebKit/UIProcess/WebFrameProxy.cpp [hunk 15]
```diff
@@ -557,6 +557,125 @@ std::optional<WebCore::PageIdentifier> WebFrameProxy::webPageIDInCurrentProcess(
...
+
+    if (m_parentFrame->m_childFrames.first().ptr() == this)
+        return nullptr;
+
+    auto it = m_parentFrame->m_childFrames.find(this);
+    if (it == m_childFrames.end()) {
+        ASSERT_NOT_REACHED();
+        return nullptr;
+    }
+    return (--it)->ptr();
```
Comment (achristensen07) [inline Source/WebKit/UIProcess/WebFrameProxy.cpp:658]: aha!  No need for rfind.  The iterator can be decremented!

**Review by achristensen07 — CHANGES_REQUESTED** [review]

**Review by achristensen07 — APPROVED** [review]

**Review by aprotyas — COMMENTED** [review]
Nice test coverage!
